### PR TITLE
Make client internally mutable

### PR DESCRIPTION
**Pull request by dani-garcia**
## 📔 Objective

We're having some deadlock problems with the Passkey API by calling certain SDK functions on the `passkey-rs` callbacks. These issues are caused because we're holding a reference to the RwLock during the execution of the Passkey operation and so calling into the SDK from inside that will deadlock.

To solve it we can't just use a lock at the client level, and so we need to move to using interior mutability, I've revived an old PR (#70) where we started doing that.

This PR is split into two commits, for ease of review: 
- [Remove all the muts and locks for uniffi](https://github.com/bitwarden/sdk/commit/c56add5c181ecd6028df3c6710da6327b2ad9166): This removes the use of RwLock in bitwarden-uniffi and the `&mut` references used in the bitwarden Client API. Note that this alone won't compile.
- [Make client internally mutable](https://github.com/bitwarden/sdk/commit/f61d10496ea5724b639ba991a7b953a7852ca34c): Implements RwLocks in Client to make it internally mutable.

Some notes of what the change entails:
- I had to wrap `EncryptionSettings` in a `RwLock<Arc<>>` to get the API working, this is quite unfortunate as it forces us to clone `EncryptionSettings` to modify it when calling `set_org_keys`. We can't wrap the `org_keys` in a `RwLock` as we will get lifetime issues in `get_key`. I think the best solution for this will be the work on creating a Crypto store that only returns opaque key references, so we might want to prioritize that.
- I extracted `external_client` from `ApiConfigurations`, as it's never mutated and it allows us to keep the same API for `get_http_client`

I hate that `Mutex` and `RwLock` return a Result in case they are poisoned, that's never our case and it makes their usage so cumbersome 😢 

Note: I've tried to change the smallest amount of code to make the change, so the secrets manager APIs (and bitwarden-json) are still mutable. If we're okay with these changes I can go back to update those in a separate PR.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation
  team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: crates/bitwarden/src/secrets_manager/secrets/update.rs [hunk 1]
```diff
@@ -44,5 +44,5 @@ pub(crate) async fn update_secret(
 
     let enc = client.get_encryption_settings()?;
```
Comment (Hinton) [inline crates/bitwarden/src/secrets_manager/secrets/update.rs:45]: I think we can reuse the same encryption setting as above?

File: crates/bitwarden/src/secrets_manager/secrets/create.rs [hunk 2]
```diff
@@ -48,5 +48,5 @@ pub(crate) async fn create_secret(
 
     let enc = client.get_encryption_settings()?;
```
Comment (Hinton) [inline crates/bitwarden/src/secrets_manager/secrets/create.rs:49]: I think we can reuse the same encryption setting above?

File: crates/bitwarden/src/secrets_manager/projects/update.rs [hunk 3]
```diff
@@ -39,5 +39,5 @@ pub(crate) async fn update_project(
 
     let enc = client.get_encryption_settings()?;
```
Comment (Hinton) [inline crates/bitwarden/src/secrets_manager/projects/update.rs:40]: I think we can reuse the same `enc` as above?

File: crates/bitwarden/src/secrets_manager/projects/create.rs [hunk 4]
```diff
@@ -40,5 +40,5 @@ pub(crate) async fn create_project(
 
     let enc = client.get_encryption_settings()?;
```
Comment (Hinton) [inline crates/bitwarden/src/secrets_manager/projects/create.rs:41]: I think we can reuse the same `enc` as above?

File: crates/bitwarden/src/mobile/tool/client_sends.rs [hunk 5]
```diff
@@ -75,11 +75,8 @@ impl<'a> ClientSends<'a> {
...
-        let key = self
-            .client
-            .get_encryption_settings()?
-            .get_key(&None)
-            .ok_or(VaultLocked)?;
+        let enc = self.client.get_encryption_settings()?;
+        let key = enc.get_key(&None).ok_or(VaultLocked)?;
         let key = Send::get_key(&send.key, key)?;
 
         let enc = buffer.encrypt_with_key(&key)?;
```
Comment (Hinton) [inline crates/bitwarden/src/mobile/tool/client_sends.rs:82]: Should we rename this now that `enc` exists above but is a different type.

File: crates/bitwarden/src/client/client.rs [hunk 6]
```diff
@@ -69,20 +69,24 @@ pub(crate) enum ServiceAccountLoginMethod {
 /// The main struct to interact with the Bitwarden SDK.
 #[derive(Debug)]
 pub struct Client {
-    token: Option<String>,
-    pub(crate) refresh_token: Option<String>,
-    pub(crate) token_expires_on: Option<i64>,
-    pub(crate) login_method: Option<LoginMethod>,
+    token: RwLock<Option<String>>,
+    pub(crate) refresh_token: RwLock<Option<String>>,
+    pub(crate) token_expires_on: RwLock<Option<i64>>,
```
Comment (Hinton) [inline crates/bitwarden/src/client/client.rs:74]: I wonder if we should find some way to group these together. You shouldn't be able to write to just one of these.

Comment (dani-garcia) [reply crates/bitwarden/src/client/client.rs]: Yeah, that's true, I extracted them to a tokens struct, which at least simplifies a bit all the locking and unlocking.

File: crates/bitwarden/src/mobile/tool/client_sends.rs (hunk 5 above)
Comment (dani-garcia) [reply crates/bitwarden/src/mobile/tool/client_sends.rs]: Good point, changed

**Review by Hinton — APPROVED** [review]

**Pull request merged**